build/commands/scripts/lit_mangler.ts

```diff
@@ -0,0 +1,200 @@
+
+        templateTags.children.push(tag)
+        return
+    }
+
+    node.forEachChild(n => {
+        getTemplateLiterals(source, n, templateTags)
+    })
+}
+
+const getLiteralsFromFile = (filepath: string) => {
+    const file = loadRaw(filepath)
+    const root: HTMLTemplateTags = { text: file.getText(), children: [], id: nextId++ }
+    getTemplateLiterals(file, file, root)
+
+    if (root.children.length === 0) {
+        throw new Error(`No templates found in ${filepath}`)
+    }
+
+    return root
+}
+
+const injectPlaceholders = (template: HTMLTemplateTags) => {
+    for (const child of template.children) {
+        const original = child.text
+        child.text = injectPlaceholders(child)
+        template.text = template.text.replace(original, `$$$$lit_mangler_${child.id}$$$$`)
+    }
+
+    return template.text
+}
+
+const replacePlaceholders = (template: HTMLTemplateTags) => {
+    for (const child of template.children) {
+        child.text = replacePlaceholders(child)
+        template.text = template.text.replaceAll(`$$lit_mangler_${child.id}$$`, child.text)
+    }
+
+    return template.text
+}
+
+let result: HTMLTemplateTags | undefined = undefined
+
+const load = (file: string) => {
+    if (result !== undefined) {
+        throw new Error('This should only be called once per file')
+    }
+
+    result = getLiteralsFromFile(file)
+    injectPlaceholders(result)
+}
+
+const write = (file: string) => {
+    if (!result) {
+        throw new Error('This should only be called after load')
+    }
+
+    const text = replacePlaceholders(result)
+    writeFileSync(file, text)
+}
+
+/**
+ * Mangles a given html template using the given mangler function.
+ *
+ * Example usage:
+ * mangle((element) => element.textContent = "foo", t => t.text.includes('allow-incognito'))
+ *
+ * @param mangler The function to use to mangle the template.
+ * @param getTemplate The template to mangle, or a predicate to find a matching template. If undefined, the first html template will be used.
+ */
+export const mangle = (mangler: (element: DocumentFragment) => void, getTemplate?: HTMLTemplateTags | ((template: HTMLTemplateTags) => boolean)) => {
+    if (!result) {
+        throw new Error('This should only be called after load!')
+    }
+
+    let template: HTMLTemplateTags | undefined = undefined
+    if (!getTemplate) {
+        template = findTemplate(() => true)
+    } else if (typeof getTemplate === 'function') {
+        template = findTemplate(getTemplate)
+    } else {
+        template = getTemplate
+    }
+
+    // If we don't have a template throw an error - mangling isn't going to work.
+    if (!template) {
+        throw new Error(`Failed to find template matching ${getTemplate}. If this was working before upstream may have changed.`)
+    }
+
+    const world = new JSDOM()
+    const document = world.window.document
+    const element = document.createElement('template')
+    element.innerHTML = template.text
```

Comment:
@thypon can you review this please

Reply:
This is server side, and build time, so looks fine.
It is also not "executed" as a site, since in `JSDOM`.

Are we sure we want to use `JSDOM` here and not a chromium-headless to interpret the DOM?
These two DOMs are not equivalent.

Reply:
I think this is okay, as per https://github.com/brave/reviews/issues/1872#issuecomment-2670119360